File: pe021/pe021.py
#!/usr/bin/python

"""
Project Euler, Problem 21:

Evaluate the sum of all the amicable numbers under 10000.

https://projecteuler.net/problem=21
"""

from math import sqrt

upper = 10000

def sum_divisors(n):
    result = 1
    for i in range(2, int(sqrt(n)) + 1):
        if n % i == 0:
            result += i
            if int(n / i) != i:
                result += int(n / i)
    return result

def main():
    divsums = [0]
    for i in range(1, upper + 1):
        divsums.append(sum_divisors(i))
    
    result = 0
    for i in range(1, len(divsums)):
        j = divsums[i]
        if j < i and divsums[j] == i:
            result += i + j
    print(result)

if __name__ == "__main__":
    main()

File: pe021/pe021.go
/*
Project Euler, Problem 21:

Evaluate the sum of all the amicable numbers under 10000.

https://projecteuler.net/problem=21
*/

package main

import (
	"fmt"
	"math"
)

const upper = 10000

func sumDivisors(n int) int {
	sum := 1
	sqrt := int(math.Sqrt(float64(n)))
	for i := 2; i <= sqrt; i++ {
		if n%i == 0 {
			sum += i
			if int(n/i) != i {
				sum += int(n / i)
			}
		}
	}
	return sum
}

func main() {
	divsums := []int{0}
	for i := 1; i <= upper; i++ {
		divsums = append(divsums, sumDivisors(i))
	}

	result := 0
	for i, d := range divsums {
		if d < i && divsums[d] == i {
			result += i + d
		}
	}
	fmt.Println(result)
}
